### PR TITLE
API/BUG: infer_dtype_from_scalar with non-nano

### LINKED ISSUE #51196: BUG: infer_dtype_from_scalar infers nanosecond dtype for second input
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [1]: from pandas.core.dtypes.cast import infer_dtype_from_scalar

In [2]: infer_dtype_from_scalar(pd.Timestamp('2020-01-01').as_unit('s'))
Out[2]: (dtype('<M8[ns]'), numpy.datetime64('2020-01-01T00:00:00.000000000'))
```

### Issue Description

nanosecond resolution is inferred for second input

### Expected Behavior

```
Out[2]: (dtype('<M8[s]'), numpy.datetime64('2020-01-01T00:00:00'))`
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : b2a26ecc538cbd04ef256b6e0de78aac700ec286
python           : 3.11.1.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.0.dev0+1430.gb2a26ecc53
numpy            : 1.25.0.dev0+446.ga701e87ce
pytz             : 2022.7
dateutil         : 2.8.2
setuptools       : 65.7.0
pip              : 22.3.1
Cython           : 0.29.33
pytest           : 7.2.0
hypothesis       : 6.52.1
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 8.8.0
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.1.0
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.11.0.dev0+1336.f5f64c5
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : 0.9.0
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None
None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #51196 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Needs whatsnew and targeted tests, opening to see if there are opinions about doing this as a bugfix or waiting to do it as an API change in 3.0

### CODE REVIEW

**Comment by jbrockmendel** [comment]
@MarcoGorelli thoughts on if/how this is worth pursuing?

**Comment by MarcoGorelli** [comment]
sorry, missed this. I'd have thought this was fine as a bug fix - I'm at jupytercon next week though so may be a bit slow to review, but will take a look when I get a chance

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
nice! just got some minor comments

File: pandas/tests/series/test_constructors.py [hunk 1]
```diff
@@ -1352,8 +1352,14 @@ def test_constructor_dict_order(self):
...
         tm.assert_series_equal(result, expected)
 
-    def test_constructor_dict_extension(self, ea_scalar_and_dtype):
+    def test_constructor_dict_extension(self, ea_scalar_and_dtype, request):
         ea_scalar, ea_dtype = ea_scalar_and_dtype
+        if isinstance(ea_scalar, Timestamp):
+            mark = pytest.mark.xfail(
+                reason="Construction from dict goes through "
+                "maybe_convert_objects which casts to nano"
+            )
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/series/test_constructors.py:1361]: is strict=True no longer the default?  if so, will update.  (also if so: yikes)

File: pandas/tests/frame/test_constructors.py [hunk 2]
```diff
@@ -1814,7 +1815,7 @@ def test_constructor_single_value(self):
     def test_constructor_with_datetimes(self):
         intname = np.dtype(np.int_).name
         floatname = np.dtype(np.float_).name
-        datetime64name = np.dtype("M8[ns]").name
+        np.dtype("M8[ns]").name
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/tests/frame/test_constructors.py:1818]: nope, will remove

File: pandas/tests/series/test_constructors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply pandas/tests/series/test_constructors.py:1361]: nvm, ignore me, it's configured to be the default

https://github.com/pandas-dev/pandas/blob/730439603919b9a38a54477f768b775f1aaf6412/pyproject.toml#L441

**Review by MarcoGorelli — APPROVED** [review]
Changes look good to me, I think it'd be OK to get this into 2.1 as it'd be a bug fix

do you want to add a whatsnew note?

**Comment by MarcoGorelli** [comment]
🤔 not sure if this might be related
```
FAILED pandas/tests/frame/methods/test_reindex.py::TestDataFrameSelectReindex::test_reindex_tzaware_fill_value - ValueError: Unexpected value for 'dtype': 'int32'. Must be 'datetime64[s]', 'datetime64[ms]', 'datetime64[us]', 'datetime64[ns]' or DatetimeTZDtype'.
```

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me

@mroeschke any comments, or good to go?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**